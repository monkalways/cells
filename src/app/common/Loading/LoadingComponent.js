import React from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, withStyles } from '@material-ui/core';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  size: PropTypes.number,
};

const defaultProps = {
  size: 100,
};

const LoadingComponent = ({ classes, size }) => (
  <div className={classes.container}>
    <CircularProgress className={classes.progress} size={size} />
  </div>
);

LoadingComponent.propTypes = propTypes;
LoadingComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  progress: {
    margin: theme.spacing.unit * 2,
  },
}))(LoadingComponent);
