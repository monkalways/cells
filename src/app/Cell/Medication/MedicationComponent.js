import _ from 'lodash';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid } from '@material-ui/core';

import CellDetaineeGrid from '../../common/CellDetaineeGrid';
import MedicationCellDetaineeCard from './MedicationCellDetaineeCard';
import MedicationFooter from './MedicationFooter';
import Loading from '../../common/Loading';

const propTypes = {
  cellDetainees: PropTypes.arrayOf(PropTypes.shape({})),
  isCellDetaineesLoaded: PropTypes.bool.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  userName: PropTypes.string,
  getCellDetainees: PropTypes.func.isRequired,
  cellName: PropTypes.string.isRequired,
  medication: PropTypes.shape({}).isRequired,
  isSavingMedication: PropTypes.bool.isRequired,
  acceptMedication: PropTypes.func.isRequired,
  rejectMedication: PropTypes.func.isRequired,
  notApplicableMedication: PropTypes.func.isRequired,
  acceptMedicationAll: PropTypes.func.isRequired,
  rejectMedicationAll: PropTypes.func.isRequired,
  notApplicableMedicationAll: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
};

const defaultProps = {
  cellDetainees: [],
  userName: null,
};

class MedicationComponent extends Component {
  componentDidMount() {
    const { cellName, getCellDetainees } = this.props;
    getCellDetainees(cellName);
  }

  getMedicationRadioButtonValue = () => {
    const { isCellDetaineesLoaded, cellDetainees, medication } = this.props;
    if (!isCellDetaineesLoaded || _.isEmpty(medication)) {
      return '';
    }
    const isAllAccept = cellDetainees
      .filter((detainee) => !detainee.location)
      .every((detainee) => medication[detainee.id].accept);
    const isAllReject = cellDetainees
      .filter((detainee) => !detainee.location)
      .every((detainee) => medication[detainee.id].reject);
    const isAllNotApplicable = cellDetainees
      .filter((detainee) => !detainee.location)
      .every((detainee) => medication[detainee.id].notApplicable);

    if (isAllAccept) return 'accept';
    if (isAllReject) return 'reject';
    if (isAllNotApplicable) return 'not-applicable';
    return '';
  };

  isSaveDisabled = () => {
    const { isCellDetaineesLoaded, cellDetainees, medication } = this.props;
    if (!isCellDetaineesLoaded && !_.isEmpty(medication)) return true;

    if (
      cellDetainees
      && cellDetainees.length > 0
      && cellDetainees.every((detainee) => detainee.location)
    ) return true;

    return false;
  };

  handleRadioGroupChange = (event) => {
    const {
      acceptMedicationAll,
      rejectMedicationAll,
      notApplicableMedicationAll,
      cellDetainees,
    } = this.props;
    const { value } = event.target;
    if (value === 'accept') acceptMedicationAll(cellDetainees.filter((detainee) => !detainee.location));
    if (value === 'reject') rejectMedicationAll(cellDetainees.filter((detainee) => !detainee.location));
    if (value === 'not-applicable') notApplicableMedicationAll(cellDetainees.filter((detainee) => !detainee.location));
  };

  handleSave = () => {
    const { medication, cellName, userName, onSave } = this.props;
    onSave(medication, cellName, userName);
  };

  render() {
    const {
      cellDetainees,
      cellName,
      isCellDetaineesLoaded,
      isAuthenticated,
      medication,
      isSavingMedication,
      acceptMedication,
      rejectMedication,
      notApplicableMedication,
    } = this.props;
    return (
      <React.Fragment>
        <CellDetaineeGrid>
          {isCellDetaineesLoaded && !isSavingMedication > 0 ? (
            <React.Fragment>
              {cellDetainees.map((cellDetainee) => (
                <Grid key={cellDetainee.id} item sm={4}>
                  <MedicationCellDetaineeCard
                    cellDetainee={cellDetainee}
                    cellName={cellName}
                    isAuthenticated={isAuthenticated}
                    medication={medication[cellDetainee.id]}
                    onAcceptClick={() => acceptMedication(cellDetainee)}
                    onRejectClick={() => rejectMedication(cellDetainee)}
                    onNotApplicableClick={() => notApplicableMedication(cellDetainee)
                    }
                  />
                </Grid>
              ))}
            </React.Fragment>
          ) : (
            <Loading />
          )}
        </CellDetaineeGrid>
        <MedicationFooter
          radioButtonValue={this.getMedicationRadioButtonValue()}
          isSavingMedication={isSavingMedication}
          isSaveDisabled={this.isSaveDisabled()}
          onRadioGroupChange={this.handleRadioGroupChange}
          onSave={this.handleSave}
        />
      </React.Fragment>
    );
  }
}

MedicationComponent.propTypes = propTypes;
MedicationComponent.defaultProps = defaultProps;

export default MedicationComponent;
