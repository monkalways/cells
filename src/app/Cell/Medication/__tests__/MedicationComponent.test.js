import React from 'react';
import { shallow } from 'enzyme';

import { MedicationComponent } from '../MedicationComponent';
import Loading from '../../../common/Loading';

describe('MedicationComponent', () => {
  let cellDetainees;
  let isCellDetaineesLoaded;
  let isAuthenticated;
  let userName;
  let cellName;
  let getCellDetainees;
  let medication;
  let isSavingMedication;
  let acceptMedication;
  let rejectMedication;
  let notApplicableMedication;
  let acceptMedicationAll;
  let rejectMedicationAll;
  let notApplicableMedicationAll;
  let onSave;

  beforeEach(() => {
    cellDetainees = [];
    isCellDetaineesLoaded = false;
    isAuthenticated = false;
    userName = null;
    cellName = 'c1';
    getCellDetainees = jest.fn();
    medication = {};
    isSavingMedication = false;
    acceptMedication = jest.fn();
    rejectMedication = jest.fn();
    notApplicableMedication = jest.fn();
    acceptMedicationAll = jest.fn();
    rejectMedicationAll = jest.fn();
    notApplicableMedicationAll = jest.fn();
    onSave = jest.fn();
  });

  const setup = () => shallow(<MedicationComponent
    cellDetainees={cellDetainees}
    isCellDetaineesLoaded={isCellDetaineesLoaded}
    isAuthenticated={isAuthenticated}
    userName={userName}
    cellName={cellName}
    getCellDetainees={getCellDetainees}
    medication={medication}
    isSavingMedication={isSavingMedication}
    acceptMedication={acceptMedication}
    rejectMedication={rejectMedication}
    notApplicableMedication={notApplicableMedication}
    acceptMedicationAll={acceptMedicationAll}
    rejectMedicationAll={rejectMedicationAll}
    notApplicableMedicationAll={notApplicableMedicationAll}
    onSave={onSave}
  />);

  it('should render cell detainee cards when cell detainees are loaded and the app is not saving medication', () => {
    cellDetainees = [
      {
        id: '123',
      },
      {
        id: '456',
      },
    ];
    medication = {
      123: {
        accept: true,
        reject: false,
        notApplicable: false,
      },
      456: {
        accept: true,
        reject: false,
        notApplicable: false,
      },
    };
    isCellDetaineesLoaded = true;
    isSavingMedication = false;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Loading component when cell detainees are not loaded', () => {
    isCellDetaineesLoaded = false;
    const wrapper = setup();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should render Loading component when cell detainees are loaded but the app is saving medication', () => {
    isCellDetaineesLoaded = true;
    isSavingMedication = true;
    const wrapper = setup();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should get cell detainees during componentDidMount', () => {
    const wrapper = setup();
    wrapper.instance().componentDidMount();

    expect(getCellDetainees).toBeCalledWith(cellName);
  });

  describe('getMedicationRadioButtonValue', () => {
    it('should return empty string if cell detainees are not loaded', () => {
      isCellDetaineesLoaded = false;

      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('');
    });

    it('should return empty string if medication data is empty', () => {
      isCellDetaineesLoaded = true;
      medication = {};

      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('');
    });

    it('should return "accept" if all in-cell detainees "accept medication"', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: null,
        },
        {
          id: '456',
          location: null,
        },
        {
          id: '789',
          location: 'Phone - In Transit',
        },
      ];
      medication = {
        123: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
        456: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
        789: {
          accept: false,
          reject: true,
          notApplicable: false,
        },
      };
      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('accept');
    });

    it('should return "reject" if all in-cell detainees "reject medication"', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: null,
        },
        {
          id: '456',
          location: null,
        },
        {
          id: '789',
          location: 'Phone - In Transit',
        },
      ];
      medication = {
        123: {
          accept: false,
          reject: true,
          notApplicable: false,
        },
        456: {
          accept: false,
          reject: true,
          notApplicable: false,
        },
        789: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
      };
      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('reject');
    });

    it('should return "not-applicable" if all in-cell detainees are "not applicable"', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: null,
        },
        {
          id: '456',
          location: null,
        },
        {
          id: '789',
          location: 'Phone - In Transit',
        },
      ];
      medication = {
        123: {
          accept: false,
          reject: false,
          notApplicable: true,
        },
        456: {
          accept: false,
          reject: false,
          notApplicable: true,
        },
        789: {
          accept: false,
          reject: true,
          notApplicable: false,
        },
      };
      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('not-applicable');
    });

    it('should return empty if some in-cell detainees "accept medication" and others "reject medication"', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: null,
        },
        {
          id: '456',
          location: null,
        },
        {
          id: '789',
          location: 'Phone - In Transit',
        },
      ];
      medication = {
        123: {
          accept: false,
          reject: true,
          notApplicable: false,
        },
        456: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
        789: {
          accept: false,
          reject: false,
          notApplicable: true,
        },
      };
      const wrapper = setup();
      const result = wrapper.instance().getMedicationRadioButtonValue();

      expect(result).toEqual('');
    });
  });

  describe('isSaveDisabled', () => {
    it('should disable save if cell detainees are not loaded', () => {
      isCellDetaineesLoaded = false;
      const wrapper = setup();

      const result = wrapper.instance().isSaveDisabled();

      expect(result).toBe(true);
    });

    it('should disable save if medication data is empty', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: null,
        },
      ];
      medication = {};
      const wrapper = setup();

      const result = wrapper.instance().isSaveDisabled();

      expect(result).toBe(true);
    });

    it('should disable save if cell detainees data is empty', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [];
      medication = {
        123: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
      };
      const wrapper = setup();

      const result = wrapper.instance().isSaveDisabled();

      expect(result).toBe(true);
    });

    it('should disable save if all cell detainees are not in cell', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: 'Phone - In Transit',
        },
      ];
      medication = {
        123: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
        456: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
      };
      const wrapper = setup();

      const result = wrapper.instance().isSaveDisabled();

      expect(result).toBe(true);
    });

    it('should enable save if some cell detainees are in cell', () => {
      isCellDetaineesLoaded = true;
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: null,
        },
      ];
      medication = {
        123: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
        456: {
          accept: true,
          reject: false,
          notApplicable: false,
        },
      };
      const wrapper = setup();

      const result = wrapper.instance().isSaveDisabled();

      expect(result).toBe(false);
    });
  });

  describe('handleRadioGroupChange', () => {
    it('should set all in-cell detainees accept medication if radio button group value is changed to "accept"', () => {
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: null,
        },
      ];
      const event = {
        target: {
          value: 'accept',
        },
      };
      const wrapper = setup();
      wrapper.instance().handleRadioGroupChange(event);

      expect(acceptMedicationAll).toBeCalledWith([
        {
          id: '456',
          location: null,
        },
      ]);
      expect(rejectMedicationAll).not.toBeCalled();
      expect(notApplicableMedicationAll).not.toBeCalled();
    });

    it('should set all in-cell detainees reject medication if radio button group value is changed to "reject"', () => {
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: null,
        },
      ];
      const event = {
        target: {
          value: 'reject',
        },
      };
      const wrapper = setup();
      wrapper.instance().handleRadioGroupChange(event);

      expect(rejectMedicationAll).toBeCalledWith([
        {
          id: '456',
          location: null,
        },
      ]);
      expect(acceptMedicationAll).not.toBeCalled();
      expect(notApplicableMedicationAll).not.toBeCalled();
    });

    it('should set all in-cell detainees not-applicable if radio button group value is changed to "not-applicable"', () => {
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: null,
        },
      ];
      const event = {
        target: {
          value: 'not-applicable',
        },
      };
      const wrapper = setup();
      wrapper.instance().handleRadioGroupChange(event);

      expect(notApplicableMedicationAll).toBeCalledWith([
        {
          id: '456',
          location: null,
        },
      ]);
      expect(acceptMedicationAll).not.toBeCalled();
      expect(rejectMedicationAll).not.toBeCalled();
    });

    it('should not set any in-cell detainees if radio button group value is changed to other values than "accept", "reject" or "not-applicable"', () => {
      cellDetainees = [
        {
          id: '123',
          location: 'Phone - In Transit',
        },
        {
          id: '456',
          location: null,
        },
      ];
      const event = {
        target: {
          value: 'other',
        },
      };
      const wrapper = setup();
      wrapper.instance().handleRadioGroupChange(event);

      expect(acceptMedicationAll).not.toBeCalled();
      expect(rejectMedicationAll).not.toBeCalled();
      expect(notApplicableMedicationAll).not.toBeCalled();
    });
  });

  it('should handle save', () => {
    const wrapper = setup();
    wrapper.instance().handleSave();

    expect(onSave).toBeCalled();
  });
});
