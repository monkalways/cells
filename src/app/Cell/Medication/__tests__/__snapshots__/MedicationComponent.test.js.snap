// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicationComponent should render cell detainee cards when cell detainees are loaded and the app is not saving medication 1`] = `
<Fragment>
  <WithStyles(CellDetaineeGridComponent)>
    <WithStyles(Grid)
      item={true}
      key="123"
      sm={4}
    >
      <WithStyles(CellDetaineeCardComponent)
        cellDetainee={
          Object {
            "id": "123",
          }
        }
        cellName="c1"
        isAuthenticated={false}
        medication={
          Object {
            "accept": true,
            "notApplicable": false,
            "reject": false,
          }
        }
        onAcceptClick={[Function]}
        onNotApplicableClick={[Function]}
        onRejectClick={[Function]}
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      key="456"
      sm={4}
    >
      <WithStyles(CellDetaineeCardComponent)
        cellDetainee={
          Object {
            "id": "456",
          }
        }
        cellName="c1"
        isAuthenticated={false}
        medication={
          Object {
            "accept": true,
            "notApplicable": false,
            "reject": false,
          }
        }
        onAcceptClick={[Function]}
        onNotApplicableClick={[Function]}
        onRejectClick={[Function]}
      />
    </WithStyles(Grid)>
  </WithStyles(CellDetaineeGridComponent)>
  <WithStyles(withRouter(MedicationFooterComponent))
    isSaveDisabled={false}
    isSavingMedication={false}
    onRadioGroupChange={[Function]}
    onSave={[Function]}
    radioButtonValue="accept"
  />
</Fragment>
`;
